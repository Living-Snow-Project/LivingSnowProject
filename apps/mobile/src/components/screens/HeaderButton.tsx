import React from "react";
import { Platform } from "react-native";
import { useColorModeValue, useTheme } from "native-base";
import { PressableOpacity } from "../PressableOpacity";
import { StockIcon } from "../media";

type IconNames =
  | "snow"
  | "save-outline"
  | "cloud-upload"
  | "settings-outline"
  | "add-circle-outline"
  | "checkmark-circle-outline";

type HeaderButtonProps = {
  testID: string;
  iconName: IconNames;
  onPress: () => void;
  placement: "left" | "right";
};

export function HeaderButton({
  testID,
  iconName,
  onPress,
  placement,
}: HeaderButtonProps) {
  const style =
    placement == "left"
      ? {
          marginLeft: 25,
        }
      : {
          marginRight: 25,
        };
  const theme = useTheme();
  const color = useColorModeValue(
    theme.colors.primary[600],
    theme.colors.primary[400]
  );

  return (
    <PressableOpacity style={style} testID={testID} onPress={onPress}>
      <StockIcon
        name={iconName}
        testID={`${testID}-${iconName}`}
        color={color}
      />
    </PressableOpacity>
  );
}
