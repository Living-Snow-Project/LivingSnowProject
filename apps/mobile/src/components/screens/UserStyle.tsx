import React from "react";
import { Avatar, Box, Text, VStack, useColorModeValue } from "native-base";
import { AlgaeRecord, isSample, recordDateFormat } from "@livingsnow/record";
import { Labels } from "../../constants";

const avatarColors = [
  "rose.300",
  "rose.600",
  "rose.900",
  "fuchsia.300",
  "fuchsia.600",
  "fuchsia.900",
  "purple.300",
  "purple.600",
  "purple.900",
  "violet.300",
  "violet.600",
  "violet.900",
  "indigo.300",
  "indigo.600",
  "indigo.900",
  "blue.300",
  "blue.600",
  "blue.900",
  "lightBlue.300",
  "lightBlue.600",
  "lightBlue.900",
  "darkBlue.300",
  "darkBlue.600",
  "darkBlue.900",
  "cyan.300",
  "cyan.600",
  "cyan.900",
  "teal.300",
  "teal.600",
  "teal.900",
  "green.300",
  "green.600",
  "green.900",
];

function computeInitials(name: string): string {
  const splitName = name.split(" ");

  // ie. John Doe = JD
  if (splitName.length > 1 && splitName[1].length != 0) {
    return `${splitName[0][0]}${splitName[1][0]}`;
  }

  // ie. John = JO or "J "
  if (name.length > 1) {
    return `${name[0]}${name[1]}`;
  }

  // fallback (Community Scientist)
  return "CS";
}

function computeColor(name: string): string {
  const hash = name
    .split("")
    .map((char) => char.charCodeAt(0))
    .reduce((current, previous) => current + previous, 0);

  return avatarColors[hash % avatarColors.length];
}

type AvatarProps = {
  color: string;
  initials: string;
};

function computeAvatarProps(name: string): AvatarProps {
  return {
    color: computeColor(name),
    initials: computeInitials(name).toUpperCase(),
  };
}

// avoid recomputing initials and color
const userAvatars: Map<string, AvatarProps> = new Map();

function getAvatarProps(name: string): AvatarProps {
  let result = userAvatars.get(name);
  if (result) {
    return { ...result };
  }

  result = computeAvatarProps(name);

  userAvatars.set(name, { ...result });

  return { ...result };
}

function getAvatar(name: string) {
  const { color, initials } = getAvatarProps(name);

  return (
    <Avatar mt={1} bg={color}>
      {initials}
    </Avatar>
  );
}

function getUserStyle(name: string | undefined, org: string | undefined) {
  const newName = name || Labels.DefaultName;

  return {
    org: org ? <Text>{org}</Text> : null,
    name: <Text fontWeight={600}>{newName}</Text>,
    avatar: getAvatar(newName),
  };
}

function getRecordInfo(record: AlgaeRecord, color: string) {
  return (
    <Text fontWeight={700} color={color}>
      {`${record.type}, ${recordDateFormat(record.date)}`}
    </Text>
  );
}

type UserStyleProps = {
  record: AlgaeRecord;
};

export function UserStyle({ record }: UserStyleProps) {
  const sample = useColorModeValue("secondary.600", "secondary.400");
  const sighting = useColorModeValue("tertiary.600", "tertiary.400");

  const color = isSample(record.type) ? sample : sighting;
  const { avatar, name, org } = getUserStyle(record.name, record.organization);

  // TODO: not ideal width % spread across 2 components, rename to TimelineHeader and include actionsMenu?
  return (
    <>
      <Box width="15%">{avatar}</Box>
      <Box width="78%" pr="1">
        <VStack ml={2}>
          {name}
          {org}
          {getRecordInfo(record, color)}
        </VStack>
      </Box>
    </>
  );
}
