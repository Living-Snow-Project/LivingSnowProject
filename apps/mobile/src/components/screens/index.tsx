export * from "./DiskUsage";
export * from "./HeaderButton";
export * from "./RecordList";
export * from "./TimelineRow";
export * from "./UserStyle";
