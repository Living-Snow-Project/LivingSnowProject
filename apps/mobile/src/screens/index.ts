export { FirstRunScreen } from "./FirstRunScreen";
export { ImagesPickerScreen } from "./ImagesPickerScreen";
export { RecordDetailsScreen } from "./RecordDetailsScreen";
export { RecordScreen } from "./RecordScreen";
export { SettingsScreen } from "./SettingsScreen";
export { TimelineScreen } from "./TimelineScreen";
