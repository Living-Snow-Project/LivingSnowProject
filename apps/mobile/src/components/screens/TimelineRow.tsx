import React from "react";
import { useNavigation } from "@react-navigation/native";
import { Box, HStack, Pressable, Text, VStack } from "native-base";
import { AlgaeRecord, Photo } from "@livingsnow/record";
import { RootStackNavigationProp } from "../../navigation/Routes";
import { Divider, ThemedBox } from "../layout";
import { UserStyle } from "./UserStyle";
import { PhotosLayout } from "../media";
import { MinimalAlgaeRecord } from "../../../types";

function bottomText({
  locationDescription,
  notes,
}: AlgaeRecord): JSX.Element[] | null {
  const result: JSX.Element[] = [];
  if (locationDescription) {
    result.push(
      <Text key={0} fontWeight="500">
        {locationDescription}
      </Text>
    );
  }

  if (notes) {
    result.push(<Text key={1}>{notes}</Text>);
  }

  return result.length > 0 ? result : null;
}

type TimelineRowProps = {
  record: AlgaeRecord;
  photos?: Photo[];
  actionsMenu?: JSX.Element;
};

export function TimelineRow({ record, photos, actionsMenu }: TimelineRowProps) {
  const { navigate } = useNavigation<RootStackNavigationProp>();
  const recordDetail: MinimalAlgaeRecord = {
    record: { ...record },
    photos: photos ? [...photos] : undefined,
  };

  return (
    <>
      <Pressable
        testID={record.id}
        onPress={() =>
          navigate("RecordDetails", { record: JSON.stringify(recordDetail) })
        }
      >
        <ThemedBox px={2} py={1}>
          <VStack>
            <HStack>
              <UserStyle record={record} />
              <Box width="7%">{actionsMenu}</Box>
            </HStack>
            {bottomText(record)}
          </VStack>
        </ThemedBox>
        <PhotosLayout photos={photos} />
      </Pressable>
      <Divider />
    </>
  );
}
