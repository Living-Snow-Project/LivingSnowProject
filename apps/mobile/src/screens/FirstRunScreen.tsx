import React from "react";
import { Box, Button, HStack, VStack, Text } from "native-base";
import { SnowIcon, UserIdentityInput } from "../components";
import { setAppSettings } from "../../AppSettings";
import { FirstRunScreenNavigationProp } from "../navigation/Routes";
import { Labels } from "../constants";

type FirstRunScreenProps = {
  navigation: FirstRunScreenNavigationProp;
};

export function FirstRunScreen({ navigation }: FirstRunScreenProps) {
  return (
    <Box px={3} mt={2}>
      <HStack alignItems="center" justifyContent="space-between">
        <SnowIcon />
        <SnowIcon />
        <VStack>
          <Text fontWeight="bold" fontSize="xl" color="pink.600">
            {Labels.LivingSnowProject}
          </Text>
          <Text color="pink.500">{Labels.Slogan}</Text>
        </VStack>
        <SnowIcon />
        <SnowIcon />
      </HStack>

      <Text my={2}>{Labels.FirstRunScreen.Usage}</Text>

      <UserIdentityInput />

      <Box alignItems="center">
        <Button
          mt={5}
          width={200}
          onPress={() => {
            setAppSettings((prev) => ({ ...prev, showFirstRun: false }));
            navigation.navigate("Timeline");
          }}
        >
          {Labels.FirstRunScreen.StartReporting}
        </Button>
      </Box>
    </Box>
  );
}
