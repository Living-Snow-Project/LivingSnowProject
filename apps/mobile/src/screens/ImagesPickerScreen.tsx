import React, { useEffect, useMemo } from "react";
import { Platform } from "react-native";
import { Ionicons } from "@expo/vector-icons";
import { Asset } from "expo-media-library";
import { useColorModeValue, useTheme } from "native-base";
import { AssetsSelector } from "expo-images-picker";
import {
  ResizeType,
  SettingsType,
  StylesType,
  CustomNavigator as CustomNavigatorType,
} from "expo-images-picker/src/Types";
import {
  ImagesPickerScreenProps,
  ImagesPickerScreenNavigationProp,
} from "../navigation/Routes";
import { HeaderButton } from "../components/screens";
import { TestIds } from "../constants";

type CustomNavigatorProps = {
  navigation: ImagesPickerScreenNavigationProp;
  onSuccess: () => void;
};

function CustomNavigator({ navigation, onSuccess }: CustomNavigatorProps) {
  useEffect(() => {
    const DoneAction = (
      <HeaderButton
        testID={TestIds.Icons.DoneSelectingPhotosIcon}
        onPress={onSuccess}
        iconName="checkmark-circle-outline"
        placement="right"
      />
    );

    /* eslint-disable react/no-unstable-nested-components */
    navigation.setOptions({
      headerRight: () => DoneAction,
    });
  }, [navigation, onSuccess]);

  return null;
}

const widgetStyles: StylesType = {
  margin: 3,
  bgColor: "white",
  spinnerColor: "blue",
  widgetWidth: 100,
  screenStyle: {
    borderRadius: 5,
    overflow: "hidden",
  },
  widgetStyle: {
    margin: 10,
  },
  videoIcon: {
    Component: Ionicons,
    iconName: "videocam",
    color: "white",
    size: 20,
  },
  selectedIcon: {
    Component: Ionicons,
    iconName: "checkmark-circle-outline",
    color: "red",
    bg: "#000000AA",
    size: 40,
  },
};

const widgetResize: ResizeType = {
  compress: 0.7,
  majorAxis: 1024,
  base64: false,
  saveTo: "jpeg",
};

// TODO: probably should do something here
const widgetErrors = {};

export function ImagesPickerScreen({
  navigation,
  route,
}: ImagesPickerScreenProps) {
  const widgetSettings: SettingsType = useMemo(
    () => ({
      getImageMetaData: false,
      initialLoad: 100,
      assetsType: ["photo"],
      minSelection: 0,
      maxSelection: 4,
      existingSelectionIds: route.params.existingSelection,
      portraitCols: 4,
      landscapeCols: 5,
    }),
    [route]
  );

  const customNavigatorWidget: CustomNavigatorType = useMemo(
    () => ({
      Component: CustomNavigator,
      props: {
        navigation,
        onSuccess: (data: Asset[]) => {
          navigation.navigate({
            name: "Record",
            params: { photos: data },
            merge: true,
          });
        },
      },
    }),
    [navigation]
  );

  const theme = useTheme();

  widgetStyles.bgColor = useColorModeValue("white", theme.colors.dark[100]);
  widgetStyles.spinnerColor = useColorModeValue(
    theme.colors.primary[600],
    theme.colors.primary[400]
  );
  /* eslint-disable prefer-destructuring */
  widgetStyles.selectedIcon.color = theme.colors.secondary[600];

  return (
    <AssetsSelector
      Settings={widgetSettings}
      Styles={widgetStyles}
      Errors={widgetErrors}
      Resize={widgetResize}
      CustomNavigator={customNavigatorWidget}
    />
  );
}
