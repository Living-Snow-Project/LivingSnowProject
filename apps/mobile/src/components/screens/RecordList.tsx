import React, { useMemo, useState } from "react";
import { useNavigation } from "@react-navigation/native";
import { Ionicons } from "@expo/vector-icons";
import { Icon, Menu, Pressable } from "native-base";
import { TimelineScreenNavigationProp } from "../../navigation/Routes";
import { Modal } from "../feedback";
import { Divider } from "../layout";
import { TimelineRow } from "./TimelineRow";
import { Labels, TestIds } from "../../constants";
import { useAlgaeRecordsContext } from "../../hooks/useAlgaeRecords";
import { productionExampleRecord } from "../../record/Record";
import { IAlgaeRecords, MinimalAlgaeRecord } from "../../../types/AlgaeRecords";

function ThreeDotsIcon() {
  return <Icon as={Ionicons} name="ellipsis-horizontal-outline" size="lg" />;
}

function ExampleRecordList() {
  const { record, photos } = productionExampleRecord();
  const records = [
    <Divider
      key={Labels.TimelineScreen.ExampleRecords}
      text={Labels.TimelineScreen.ExampleRecords}
    />,
    <TimelineRow key="single_example" record={record} photos={photos} />,
  ];

  return (
    <>
      {records[0]}
      {records[1]}
    </>
  );
}

function useDownloadedRecordList() {
  const algaeRecords = useAlgaeRecordsContext();
  const records = algaeRecords.getDownloaded();
  const pendingLength = algaeRecords.getPending().length;

  const renderRecords = useMemo(() => {
    const result: JSX.Element[] = [];

    // "Downloaded Divider" is only useful if there are pending records
    if (pendingLength > 0) {
      result.push(
        <Divider
          key={Labels.TimelineScreen.DownloadedRecords}
          text={Labels.TimelineScreen.DownloadedRecords}
        />,
      );
    }

    records.forEach((record) =>
      result.push(
        <TimelineRow
          key={record.id}
          record={record}
          photos={record.photos.appPhotos}
        />,
      ),
    );

    return result;
  }, [records, pendingLength]);

  if (records.length == 0) {
    return null;
  }

  return renderRecords;
}

type PendingTimelineRowProps = {
  localRecord: MinimalAlgaeRecord;
  algaeRecords: IAlgaeRecords;
};

function PendingTimelineRow({
  localRecord,
  algaeRecords,
}: PendingTimelineRowProps) {
  const [isOpen, setIsOpen] = useState(false);
  const navigation = useNavigation<TimelineScreenNavigationProp>();

  const onPressDelete = () => setIsOpen(true);
  const onConfirmDelete = () => algaeRecords.delete(localRecord.record.id);

  const onEdit = () =>
    navigation.navigate("Record", {
      record: JSON.stringify(localRecord.record),
    });

  const menuTrigger = (props: any) => (
    <Pressable {...props} testID={TestIds.RecordList.MenuTrigger}>
      <ThreeDotsIcon />
    </Pressable>
  );

  const actionsMenu = (
    <>
      <Modal
        body="Are you sure you want to delete this pending record?"
        header="Confirm Delete"
        isOpen={isOpen}
        testId={TestIds.Modal.DeletePendingRecord}
        setIsOpen={setIsOpen}
        onConfirm={onConfirmDelete}
      />
      <Menu trigger={menuTrigger}>
        <Menu.Item
          onPress={onEdit}
          testID={TestIds.RecordList.EditRecordAction}
        >
          Edit
        </Menu.Item>
        <Menu.Item
          onPress={onPressDelete}
          testID={TestIds.RecordList.DeleteRecordAction}
        >
          Delete
        </Menu.Item>
      </Menu>
    </>
  );

  return (
    <TimelineRow
      record={localRecord.record}
      photos={localRecord.photos}
      actionsMenu={actionsMenu}
    />
  );
}

function usePendingRecordList() {
  const algaeRecords = useAlgaeRecordsContext();
  const pendingRecords = algaeRecords.getPending();

  const renderRecords = useMemo(() => {
    const result: JSX.Element[] = [];

    if (pendingRecords.length == 0) {
      return null;
    }

    result.push(
      <Divider
        key={Labels.TimelineScreen.PendingRecords}
        text={Labels.TimelineScreen.PendingRecords}
      />,
    );

    pendingRecords.forEach((record, index) => {
      /* eslint-disable react/no-array-index-key */
      result.push(
        <PendingTimelineRow
          key={`pending-${index}`}
          localRecord={record}
          algaeRecords={algaeRecords}
        />,
      );
    });

    return result;
  }, [algaeRecords, pendingRecords]);

  return renderRecords;
}

export { ExampleRecordList, useDownloadedRecordList, usePendingRecordList };
