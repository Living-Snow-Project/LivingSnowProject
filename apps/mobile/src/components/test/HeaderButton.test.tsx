import React from "react";
import { Platform } from "react-native";
import { fireEvent, render } from "@testing-library/react-native";
import { NativeBaseProviderForTesting } from "../../../jesttest.setup";
import { HeaderButton } from "../screens";

describe("HeaderButton test suite", () => {
  const testID = "test-button";

  test("renders left on ios", () => {
    const onPress = jest.fn();
    const { getByTestId, toJSON } = render(
      <NativeBaseProviderForTesting>
        <HeaderButton
          testID={testID}
          iconName="snow"
          onPress={onPress}
          placement="left"
        />
      </NativeBaseProviderForTesting>
    );

    fireEvent.press(getByTestId(testID));
    expect(onPress).toBeCalled();
    expect(toJSON()).toMatchSnapshot();
  });

  test("renders right on android", () => {
    Platform.OS = "android";
    const onPress = jest.fn();
    const { getByTestId, toJSON } = render(
      <NativeBaseProviderForTesting>
        <HeaderButton
          testID={testID}
          iconName="snow"
          onPress={onPress}
          placement="right"
        />
      </NativeBaseProviderForTesting>
    );

    Platform.OS = "ios";

    fireEvent.press(getByTestId(testID));
    expect(onPress).toBeCalled();
    expect(toJSON()).toMatchSnapshot();
  });
});
