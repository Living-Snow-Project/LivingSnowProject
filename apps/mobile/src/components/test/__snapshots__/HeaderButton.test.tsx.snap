// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderButton test suite renders left on ios 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    accessible={true}
    collapsable={false}
    dataSet={Object {}}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "marginLeft": 24,
      }
    }
    testID="test-button"
  >
    <Text />
  </View>
</RNCSafeAreaProvider>
`;

exports[`HeaderButton test suite renders right on android 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    accessible={true}
    collapsable={false}
    dataSet={Object {}}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "marginRight": 24,
      }
    }
    testID="test-button"
  >
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        Array [
          Object {
            "color": "#0891b2",
            "fontSize": 32,
          },
          undefined,
          Object {
            "fontFamily": "ionicons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
      testID="snow"
    >
      ï”¬
    </Text>
  </View>
</RNCSafeAreaProvider>
`;
